### PR TITLE
Endpoints and Evals

### PR DESCRIPTION
**Pull request by zknill**
Add endpoint and eval commands on the CLI

```
unweave endpoint new <exec-id>
unweave eval new <exec-id>
unweave endpoint attach-eval <endpoint-id> <eval-id>
unweave endpoint check
unweave endpoint check-stauts <check-id>
```

### CODE REVIEW

File: client/endpoints.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/unweave/cli/ui"
+	"github.com/unweave/unweave/api/types"
+)
+
+type EndpointService struct {
+	client *Client
+}
+
+func (s *EndpointService) List(ctx context.Context, userID, projectID string) ([]types.Endpoint, error) {
+	uri := fmt.Sprintf("projects/%s/%s/endpoints", userID, projectID)
```
Comment (ernesto-jimenez) [inline client/endpoints.go:16]: Question: do we have too hard-code URLs in both the client and the server? How do we ensure they are aligned?

File: cmd/endpoints.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	endpoints, err := uwc.Endpoints.List(ctx, owner, projectName)
+	if err != nil {
+		return err
+	}
+
+	for _, endpoint := range endpoints {
+		ui.Infof("id: %s", endpoint.ID)
+		ui.Infof("exec id: %s", endpoint.ExecID)
+		ui.Infof("http: %s\n", endpoint.HTTPEndpoint)
+	}
```
Comment (ernesto-jimenez) [inline cmd/endpoints.go:42]: 💡 tangential idea for the future: we'll probably end-up wanting to provide output in structured formats like JSON.

One small improvement in the future would probably be implementing something like `ui.Output(endpoints)` and handle the presentation in text/json/something else.

File: main.go [hunk 3]
```diff
@@ -405,6 +405,103 @@ func init() {
...
+	endpointNewCmd := &cobra.Command{
+		Use:   "new <exec-id>",
+		Short: "Create a new endpoint in Unweave",
+		Long: wordwrap.String("Create a new endpoint in Unweave.\n\n"+
+			"Eg. unweave endpoint new <exec-id>\n\n",
+			ui.MaxOutputLineLength),
+		Args:    cobra.ExactArgs(1),
+		Aliases: []string{"new", "n", "create", "c"},
+		RunE:    cmd.EndpointCreate,
+	}
```
Comment (noorvir) [inline main.go:426]: This is a wip syntax right? I think the way we want to expose this is probably: 
`unweave endpoint new <endpoint-name>` and then `unweave deploy <endpoint-name>`.

**Review by noorvir — APPROVED** [review]

**Review by ernesto-jimenez — COMMENTED** [review]
Sorry for the delay. Code-wise everything looked OK, but I wanted to test things and have been struggling a bit with staging being inaccessible and how to connect to the local env.

File: main.go (hunk 3 above)
Comment (zknill) [reply main.go:426]: Yes, likely. Once you've got the ERD for this we can hash out the details, and I can make the changes easily :100:

File: client/endpoints.go (hunk 1 above)
Comment (zknill) [reply client/endpoints.go:16]: Yes, the paths are 'hard coded' in the client (cli), and the router's mux. 
Right now, we make sure they are aligned by smoke testing / using them

File: cmd/endpoints.go (hunk 2 above)
Comment (zknill) [reply cmd/endpoints.go:42]: This is _VERY_ much a temporary output, to have something, while we work out what's the best something we can print.

**Pull request merged**